Cloud ObjectStore

VJalili opened the pull request:
_This PR is a follow up from #4352 which we had to close it due to an issue with rebase. You may check that PR for some comments and reviews._ 

The `Cloud` ObjectStore is a provider-agnostic cloud-based storage. In other words, the `Cloud` ObjectStore can read/write from various cloud-based storages without a provider-specific implementation; it leverages on [`CloudBridge`](https://github.com/gvlproject/cloudbridge) which provides a unified access to various providers. 

This PR sets a base for the User-based ObjectStore we have been working on. Initially, the `Cloud` ObjectStore was part of the User-based ObjectStore (PR [#4314](https://github.com/galaxyproject/galaxy/pull/4314)), and it is being submitted as a separate PR following the comments on the PR [#4314](https://github.com/galaxyproject/galaxy/pull/4314).

To simplify the comparison with `S3ObjectStore`, the `Cloud` ObjectStore provides the very same functionality as `S3ObjectStore`, i.e., read/write to AWS S3 buckets. The only difference is: `S3ObjectStore` leverages on `boto` while `Cloud` uses `CloudBridge`.

VJalili committed:
Added CloudBridge to the requirements list.

VJalili committed:
Added CloudBridge to the dependencies check.

VJalili committed:
(1) Added Cloud to ObjectStore; (2) updated CloudBridge dependencies.

mvdbeek committed:
Merge remote-tracking branch 'origin/release_17.05'

VJalili committed:
- Updated `Cloud` to account for changes in `CloudBridge` interface.
- Added `Cloud` to objectstore import.

VJalili committed:
Updated `Babel` and `requests` packages version to meet the minimum
requirements of `cloudbridge` package.

VJalili committed:
Reverted the changes on package versions, as these changes are
requests via a different PR.

VJalili committed:
Merge remote-tracking branch 'upstream/master'

VJalili committed:
Changed boto import style.

VJalili committed:
Updated CloudBridge to its current latest version.

VJalili committed:
Commented on boto import, and removed a `use_reduced_redundancy` config.

VJalili committed:
Removed boto import, and replaced all S3-specific exception catches with
a generic exception--a temporary solution till CloudBridge wraps the
exceptions properly.

VJalili committed:
Consolidated two upload methods, i.e., upload from file and string,
because CloudBridge internally maps to appropriate functions depending
on the input type.

VJalili committed:
Removed a S3-related comment, and swapped an if/else condition.

VJalili committed:
Removed a comment section.

VJalili committed:
Changed indentation of two lines.

VJalili committed:
Update CloudBridge to its current latest

Should use the wheel created at [this PR](https://github.com/galaxyproject/starforge/pull/139).

VJalili committed:
Add Cloud configuration to object_store_conf.xml.sample

VJalili committed:
Merge remote-tracking branch 'remotes/upstream/dev' into CloudObjectStore

VJalili committed:
Removed white spaces.

VJalili committed:
Updated requirements: sqlalchemy-migrate and pbr.

VJalili committed:
Removed a white space between parenthesis and arguments

VJalili committed:
Surrounded CloudBridge import in a try-catch block.

VJalili commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,568 @@
+                log.debug("Cache cleaning done. Total space freed: %s", convert_bytes(deleted_amount))
+                return
+
+    def _get_bucket(self, bucket_name):
```
@nuwang; referring to your comment on #4352; the logic here is "get the bucket, if not available, then create it first, then return the bucket". However, the cloudbridge logic is "get the bucket, if not available, return none" ([see this line](https://github.com/VJalili/cloudbridge/blob/master/cloudbridge/cloud/providers/aws/resources.py#L884)). So, do you want to update cloudbridge and have this logic implemented internally?

nuwang replied:
@VJalili @afgane I think that sequence is alright. I was thinking more about the outer loop - having to loop 5 times before getting a bucket. Calling cloudbridge.create() should guarantee that the bucket is available when it returns, or it should raise an exception. Having to loop suggests that the cloudbridge code is not portable across providers (e.g. this might work in one go on openstack, and a developer would think the code is working correctly, but then things will actually fall over on AWS). We could either document this as being necessary at the interface level, or handle this internally within cloudbridge. Do you know when specifically this occurs?

nuwang replied:
I've created an issue here so we can continue the discussion: https://github.com/gvlproject/cloudbridge/issues/64

VJalili replied:
@nuwang I agree. I removed the retries here, and now we rely on the first try :)

nuwang replied:
@VJalili Ok, sounds good :-) Will close the corresponding issue.

afgane commented:
Looks good to me as an intermediate step that focuses on replacing AWS/`boto` code with the cloud-agnostic `CloudBridge` code. Much of this will get redone again with the move toward the user-based object store so even though there are a few AWS-specific terms, that's fine.

+1

jmchilton committed:
Rework conditional dependency handling in galaxy.objectstore.cloud.

This file may be loaded (e.g. for testing) but not used, so we shouldn't log a generic error about cloudbridge being unavailable, if someone attempts to actually use the object store and it isn't available then raise an informative exception.

jmchilton commented:
I've opened a PR to rework the conditional dependencies here (https://github.com/VJalili/galaxy/pull/3) - see PR for more information.

I'd also like to see this line:

```
logging.getLogger('boto').setLevel(logging.INFO)  # Otherwise boto is quite noisy
```

Moved somewhere else - unconditionally applying it as a result of importing a random module in Galaxy seems wrong - but the same line appears in s3.py so I wouldn't hold up this PR on that I guess.

VJalili committed:
Merge pull request #3 from jmchilton/CloudObjectStore

Rework conditional dependency handling in galaxy.objectstore.cloud.

VJalili committed:
Updated an error message.

VJalili committed:
Removed `boto` logging level set.

VJalili commented:
>     logging.getLogger('boto').setLevel(logging.INFO)  # Otherwise boto is quite noisy
> Moved somewhere else - unconditionally applying it as a result of importing a random module in Galaxy seems wrong - but the same line appears in s3.py so I wouldn't hold up this PR on that I guess.

@jmchilton Thanks for mentioning this. Actually, this line should be removed as we're not (explicitly) using `boto`; I removed it.

VJalili committed:
Removed retries for getting a bucket.

natefoo commented:
I've had about 2 minutes to look at this and have about 1 minute to write this comment but I wanted to get something in here just so I'll get notifications as a participant. ;)

Without having looked myself, it seems? like there's a lot of duplication from S3? Can these not be refactored into a CachingObjectStore (which I could've sworn we already had as a parent of the S3ObjectStore)?

VJalili commented:
@natefoo you're right, the `Cloud` class is very close to the `S3ObjectStore` ObjectStore, and that is intentional :) the reason is #4314. 

The `Cloud` module introduced in this PR, will be updated in follow-up PRs to enable a per user-based ObjectStore (i.e., plug-and-play your own resource such as Amazon S3, Microsoft Azure, Google drive, and etc.).

natefoo commented:
Ah, so the plan is to merge them in the future.

VJalili commented:
Actually, I would also prefer to merge the overlapping modules in favor of orthogonal functionality. But as you may read in #4314, it seems it is more "galaxy-ish" to keep S3 and Azure (even-though you get the very same functionality with `Cloud`) for backward compatibility. Therefore, merging `cloud` with `s3` and `azure_blob` is not in the plan :)

natefoo commented:
In that case, can we aim for refactoring the duplication into a caching object store class and have S3, Cloud, and Azure subclass from that?

dannon commented:
@natefoo That was the plan AFAIK, we just wanted to keep the initial implementation of this simple and isolated, and make a refactoring pass after.

@VJalili Sorry for the delay getting to this, I'll test it for ya today and see how it goes.

dannon commented:
In testing, I'm finding that the *path* to the cached version of a file is actually what gets uploaded as the contents of the file, instead of the contents as expected.

I set up a new IAM user, uploaded a new dataset, and observed that in my S3 bucket the contents of the uploaded file were actually something like `/Users/yoplait/work/galaxy/database/object_store_cache/000/dataset_183.dat`, instead of the data expected.

The cached file on disk is fine, but when I *delete* that cached file, the file is replaced by a plain text file containing the string as mentioned above.  That's a good news / bad news thing in that the good news is that fetching the file from S3 is working, but we definitely need to fix the contents that get uploaded.

Feel free to move it back to 17.09 if you feel inclined, but because this is a fairly critical issue for this PR (and I'm not seeing this as an urgent feature for 17.09), I'm going to bump this to 18.01.

VJalili committed:
Resolved a bug with uploading a dataset using Cloud (not differentiating
between uploading from a file vs. string).

VJalili commented:
Interesting point, thanks @dannon
I fixed it. 

It seems there was a piece of code in the [original PR](https://github.com/galaxyproject/galaxy/pull/4314/files#diff-97a86fece2684b3b78a5b500985594acR389) which was differentiating between uploading from a file and object. The same logic was moved to this PR, but it seems I removed it at [this commit](https://github.com/galaxyproject/galaxy/pull/4487/commits/fafb019f4be3b33ad9cab4bfd2ded6788b419592) assuming it is handled internally by CloudBridge; maybe a misunderstanding. I reverted the commit.

VJalili committed:
Removed an unused variable (i.e., mb_size) from Cloud

dannon commented:
Merged locally resolving conflicts.  Thanks for the fixes @VJalili, I confirmed that the dataset contents bug above was fixed and stuff seemed to be working pretty well to me.

VJalili commented:
Thanks @dannon .
